extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ContextRegistrarBuildItem.java

```diff
@@ -1,20 +1,56 @@
 package io.quarkus.arc.deployment;
 
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.jboss.jandex.DotName;
+
 import io.quarkus.arc.processor.ContextRegistrar;
 import io.quarkus.builder.item.MultiBuildItem;
 
 /**
- * Make it possible to register a custom CDI context.
+ * Make it possible to register one or more custom CDI contexts.
+ * If you are registering a new context, you should also pass in the annotation value into the constructor either in
+ * {@link DotName} form, or as {@code Class<? extends Annotation>}.
+ *
+ * This information is then leveraged in {@link CustomScopeAnnotationsBuildItem} which allows consumers to browse
+ * all known custom scoped within deployment even early in the build process.
  */
 public final class ContextRegistrarBuildItem extends MultiBuildItem {
 
     private final ContextRegistrar contextRegistrar;
+    private final Collection<DotName> annotationNames;
 
-    public ContextRegistrarBuildItem(ContextRegistrar contextRegistrar) {
+    public ContextRegistrarBuildItem(ContextRegistrar contextRegistrar, DotName... annotationsNames) {
         this.contextRegistrar = contextRegistrar;
+        if (annotationsNames == null) {
+            throw new IllegalArgumentException("Parameter annotationNames cannot be null!");
+        }
+        Collection<DotName> names = new ArrayList<>(annotationsNames.length);
+        for (DotName name : annotationsNames) {
+            names.add(name);
+        }
+        this.annotationNames = names;
+    }
+
+    public ContextRegistrarBuildItem(ContextRegistrar contextRegistrar, Class<? extends Annotation>... annotationsClasses) {
+        this.contextRegistrar = contextRegistrar;
```

Comment:
You could call the `ContextRegistrarBuildItem(ContextRegistrar contextRegistrar, DotName... annotationsNames)` constructor instead... something like:

```java
this(contextRegistrar, annotationsClasses.stream().map(Class::getName).map(DotName::createSimple).toArray(DotName[]::new));
```